### PR TITLE
[analyzer] Add CTU On-demand analysis support

### PR DESCRIPTION
**Pull request by gamesh411**
On-demand parsing of the ASTs can be now enabled if the relevant Phabricator patch lands. If enabled, on-the-fly analysis removes the need to create AST dumps, but in turn the compilation database must be provided to Clang. The CTU Index also changes if the option is enabled. The mapping form `function name (USR)` -> `AST dump file` becomes `function name (USR)` -> `path to source file`.

### CODE REVIEW

**Comment by gamesh411** [comment]
Do note, that this modification depends on the clang801 branch getting the pull request for supporting on-demand-parsing. This is currently not merged, so the ci may fail.

**Comment by gamesh411** [comment]
Renamed the feature to match the naming used in Clang.

File: analyzer/codechecker_analyzer/analyzers/clangsa/analyzer.py [hunk 1]
```diff
@@ -227,6 +227,15 @@ def construct_analyzer_cmd(self, result_handler):
                 if ctu_display_progress:
                     analyzer_cmd.extend(ctu_display_progress)
 
+                if config.ctu_on_demand:
+                    analyzer_cmd.extend(
+                        ['-Xclang', '-analyzer-config', '-Xclang',
+                         'ctu-on-demand-parsing=true', '-Xclang',
```
[the remark this answers is not in the document]
Comment (gyorb) [reply analyzer/codechecker_analyzer/analyzers/clangsa/analyzer.py]: Shouldn't this be clang version dependent? When will it be available in clang?

File: analyzer/codechecker_analyzer/analyzers/clangsa/ctu_manager.py [hunk 2]
```diff
@@ -120,27 +120,47 @@ def generate_ast(triple_arch, action, source, config, env):
...
+        mapped_path = path if ctu_on_demand else ast_dump_path(path)
+
+        func_ast_list.append(mangled_name + " " + mapped_path)
     return func_ast_list
 
 
 def map_functions(triple_arch, action, source, config, env,
                   func_map_cmd, temp_fnmap_folder):
-    """ Generate function map file for the current source. """
+    """ Generate function map file for the current source.
```
Comment (gyorb) [inline analyzer/codechecker_analyzer/analyzers/clangsa/ctu_manager.py]: Do we use the same file name for both type of data? Would there be any conflict if I switch between the two types and the original ctu tries to read up the function map file created for the on-demand?

File: analyzer/codechecker_analyzer/cmd/analyze.py [hunk 3]
```diff
@@ -345,6 +345,16 @@ def add_arguments_to_parser(parser):
                                     "'<OUTPUT_DIR>/ctu-dir'. (These files "
                                     "will not be cleaned up in this mode.)")
 
+        ctu_opts.add_argument('--ctu-on-demand',
```
Comment (gyorb) [inline analyzer/codechecker_analyzer/cmd/analyze.py]: Does the user get any warnings if `--ctu --ctu-collect --ctu-on-demand` is used together?

**Comment by gyorb** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- analyzer/codechecker_analyzer/analyzers/clangsa/analyzer.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/CodeChecker/codechecker/pullRequest?prid=3892752&bid=13556301)

**Comment by gyorb** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- analyzer/codechecker_analyzer/analyzers/clangsa/analyzer.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/CodeChecker/codechecker/pullRequest?prid=3892752&bid=13556301)

**Comment by gyorb** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- analyzer/tests/functional/ctu_failure/test_ctu_failure.py  1
- analyzer/codechecker_analyzer/analyzers/clangsa/analyzer.py  1
         

Clones added
============
- analyzer/tests/functional/ctu_failure/test_ctu_failure.py  28
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/CodeChecker/codechecker/pullRequest?prid=3892752&bid=13556301)

File: analyzer/codechecker_analyzer/cmd/analyze.py (hunk 3 above)
Comment (gamesh411) [reply analyzer/codechecker_analyzer/cmd/analyze.py:460]: I have created a new option-group called ast-loading-mode, that way it will not conflict with the original ctu-options.

File: analyzer/codechecker_analyzer/pre_analysis_manager.py [hunk 4]
```diff
@@ -115,8 +115,16 @@ def pre_analyze(params):
                 ctu_triple_arch.get_triple_arch(action, action.source,
                                                 config,
                                                 analyzer_environment)
-            ctu_manager.generate_ast(triple_arch, action, action.source,
-                                     config, analyzer_environment)
+
+            LOG.debug("config", config)
```
[the remark this answers is not in the document]
Comment (gamesh411) [reply analyzer/codechecker_analyzer/pre_analysis_manager.py]: Debug line left in, fixed it.

File: analyzer/codechecker_analyzer/analyzers/clangsa/analyzer.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (gamesh411) [reply analyzer/codechecker_analyzer/analyzers/clangsa/analyzer.py]: Feature detection decides now whether the option exists for the user or not.

File: analyzer/codechecker_analyzer/analyzers/clangsa/ctu_manager.py (hunk 2 above)
Comment (gamesh411) [reply analyzer/codechecker_analyzer/analyzers/clangsa/ctu_manager.py:190]: Right now the index format is (usr-name -> file-system-path). In case of dump based (vanilla) ctu the path is something that has an `.ast` postfix. If it has any other postfix, the entry is supposed to be used with on-demand loader.

File: docs/analyzer/user_guide.md [hunk 5]
```diff
@@ -645,6 +645,7 @@ usage: CodeChecker analyze [-h] [-j JOBS]
                            [--tidyargs TIDY_ARGS_CFG_FILE]
                            [--tidy-config TIDY_CONFIG] [--timeout TIMEOUT]
                            [--ctu | --ctu-collect | --ctu-analyze]
+                           [--ctu-on-demand]
```
Comment (csordasmarton) [inline docs/analyzer/user_guide.md:648]: I think this option was renamed to `--ctu-ast-loading-mode`

File: docs/analyzer/user_guide.md [hunk 6]
```diff
@@ -1243,7 +1244,7 @@ cross translation unit analysis arguments:
                         analysis, using already available extra files in
                         '<OUTPUT_DIR>/ctu-dir'. (These files will not be
                         cleaned up in this mode.)
-  --ctu-on-the-fly      If specified, the 'collect' phase will not create the
+  --ctu-on-demand       If specified, the 'collect' phase will not create the
```
Comment (csordasmarton) [inline docs/analyzer/user_guide.md:1247]: Update the docs with the `--ctu-ast-loading-mode` help.

Comment (gamesh411) [reply docs/analyzer/user_guide.md]: Thanks for catching these! fixed the docs 📖 .

**Comment by gyorb** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- analyzer/tests/functional/ctu_failure/test_ctu_failure.py  1
- analyzer/codechecker_analyzer/analyzers/clangsa/analyzer.py  1
         

Clones added
============
- analyzer/tests/functional/ctu_failure/test_ctu_failure.py  28
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/CodeChecker/codechecker/pullRequest?prid=3892752&bid=13556301)

**Comment by gyorb** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- analyzer/codechecker_analyzer/analyzers/clangsa/analyzer.py  1
- analyzer/tests/functional/ctu_failure/test_ctu_failure.py  1
         

Clones added
============
- analyzer/tests/functional/ctu_failure/test_ctu_failure.py  25
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/CodeChecker/codechecker/pullRequest?prid=3892752&bid=13556301)

File: analyzer/codechecker_analyzer/analyzers/clangsa/ctu_manager.py [hunk 7]
```diff
@@ -95,8 +97,39 @@ def merge_ctu_func_maps(ctu_dir, ctu_func_map_file, ctu_temp_fnmap_folder):
...
+    triple_arch_dir = Path(config.ctu_dir, triple_arch)
+    # Ensure, that architecture directory exists.
+    triple_arch_dir.mkdir(parents=True, exist_ok=True)
+
+    invocation_list = triple_arch_dir / 'invocation-list.yml'
+
+    source_path = Path(source).resolve()
+
+    cmd = ctu_triple_arch.get_compile_command(action, config, source)
+    # __clang__analyzer__ macro needs to be set in the imported TUs too.
```
Comment (martong) [inline analyzer/codechecker_analyzer/analyzers/clangsa/ctu_manager.py:115]: This is a good observation, I am not sure if we set this in the vanilla CTU mode at all.

File: analyzer/tests/functional/ctu_failure/test_ctu_failure.py [hunk 8]
```diff
@@ -209,7 +340,39 @@ def test_ctu_fallback(self):
...
-        output = self.__do_ctu_all(reparse=False,
+        output = self.__do_ctu_all(on_demand=False,
+                                   extra_args=[
+                                       "--verbose", "debug",
+                                       "-e", "debug.ExprInspection",
+                                       "--ctu-reanalyze-on-failure"
+                                   ])
+
+        # lib.c should be logged as its AST is loaded by Clang
+        self.assertRegex(output, r"CTU loaded AST file: .*lib\.c.ast")
```
Comment (martong) [inline analyzer/tests/functional/ctu_failure/test_ctu_failure.py:351]: Does Clang really log the `.ast` postfix? That could be misleading when we have the on-demand parsing.

Comment (martong) [reply analyzer/tests/functional/ctu_failure/test_ctu_failure.py:351]: I checked it, Clang just writes out the file name it loaded and in case of on-demand that should be a `.c` or `.cpp` file, isn't that so?

File: analyzer/codechecker_analyzer/analyzers/clangsa/ctu_manager.py (hunk 7 above)
Comment (gamesh411) [reply analyzer/codechecker_analyzer/analyzers/clangsa/ctu_manager.py:115]: We do set it with ast-dump-based mode, but we do not if we do it manually (ctu without CodeChecker). Frankly, I did not even know this was a requirement... :astonished:

File: analyzer/tests/functional/ctu_failure/test_ctu_failure.py (hunk 8 above)
Comment (gamesh411) [reply analyzer/tests/functional/ctu_failure/test_ctu_failure.py]: Fixed this, thanks for catching! These tests are not run, because we do not have a clang that has the feature **and** otherwise passes every test as well. Would came out during the update to the next supporting Clang version, but it is better to be proactive. :+1:

**Comment by gyorb** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- analyzer/tests/functional/ctu_failure/test_ctu_failure.py  1
- analyzer/codechecker_analyzer/analyzers/clangsa/analyzer.py  1
         

Clones added
============
- analyzer/tests/functional/ctu_failure/test_ctu_failure.py  25
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/CodeChecker/codechecker/pullRequest?prid=3892752&bid=13556301)

**Comment by gyorb** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- analyzer/codechecker_analyzer/analyzers/clangsa/analyzer.py  1
- analyzer/tests/functional/ctu_failure/test_ctu_failure.py  1
         

Clones added
============
- analyzer/tests/functional/ctu_failure/test_ctu_failure.py  25
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/CodeChecker/codechecker/pullRequest?prid=3892752&bid=13556301)

**Comment by gyorb** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- analyzer/tests/functional/ctu_failure/test_ctu_failure.py  1
- analyzer/codechecker_analyzer/analyzers/clangsa/analyzer.py  1
         

Clones added
============
- analyzer/tests/functional/ctu_failure/test_ctu_failure.py  25
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/CodeChecker/codechecker/pullRequest?prid=3892752&bid=13556301)

File: analyzer/tests/functional/ctu_failure/test_ctu_failure.py (hunk 8 above)
Comment (gamesh411) [reply analyzer/tests/functional/ctu_failure/test_ctu_failure.py]: Finally fixed those as well, thanks 👍

**Comment by gyorb** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- analyzer/codechecker_analyzer/analyzers/clangsa/analyzer.py  1
- analyzer/tests/functional/ctu_failure/test_ctu_failure.py  1
         

Clones added
============
- analyzer/tests/functional/ctu_failure/test_ctu_failure.py  25
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/CodeChecker/codechecker/pullRequest?prid=3892752&bid=13556301)

File: analyzer/tests/functional/ctu_failure/test_ctu_failure.py (hunk 8 above)
Comment (martong) [reply analyzer/tests/functional/ctu_failure/test_ctu_failure.py]: :+1:

File: analyzer/codechecker_analyzer/analyzers/clangsa/ctu_manager.py [hunk 9]
```diff
@@ -95,8 +97,39 @@ def merge_ctu_func_maps(ctu_dir, ctu_func_map_file, ctu_temp_fnmap_folder):
...
+    cmd.extend(['-D__clang_analyzer__', '-w'])
+
+    # Dump the mapping as a JSON dictionary, then strip the braces.
+    # The resulting line is a proper YAML mapping entry.
+    # TODO: Use proper YAML serialization
```
Comment (gyorb) [inline analyzer/codechecker_analyzer/analyzers/clangsa/ctu_manager.py:120]: the yaml library will be introduced in another PR too so I think we can use that to generate proper yaml files.

File: analyzer/codechecker_analyzer/analyzers/clangsa/ctu_manager.py [hunk 10]
```diff
@@ -95,8 +97,39 @@ def merge_ctu_func_maps(ctu_dir, ctu_func_map_file, ctu_temp_fnmap_folder):
...
+    invocation_line = json.dumps({str(source_path): cmd})[1:-1]
+
+    LOG.debug_analyzer("Appending invocation list item '%s'", invocation_line)
+
+    # Append the next entry into the invocation list yaml.
+    with invocation_list.open('a') as invocation_file:
```
Comment (gyorb) [inline analyzer/codechecker_analyzer/analyzers/clangsa/ctu_manager.py:126]: please set the encoding and errors arguments for the file open so valid utf-8 files are created even if the host machine uses some other encoding.
```suggestion
    with invocation_list.open('a', encoding='utf-8', errors='ignore') as invocation_file:
```

File: analyzer/codechecker_analyzer/cmd/analyze.py [hunk 11]
```diff
@@ -466,6 +466,25 @@ def add_arguments_to_parser(parser):
                                    "same translation unit without "
                                    "Cross-TU enabled.")
 
+        # Only check for AST loading modes if CTU is available.
+        if analyzer_types.is_ctu_on_demand_available(context):
+            ctu_opts.add_argument('--ctu-ast-creation',
```
Comment (gyorb) [inline analyzer/codechecker_analyzer/cmd/analyze.py:471]:
```suggestion
            ctu_opts.add_argument('--ctu-ast-mode',
```
what do you think?

File: analyzer/codechecker_analyzer/cmd/analyze.py [hunk 12]
```diff
@@ -466,6 +466,25 @@ def add_arguments_to_parser(parser):
...
+                                  help="Choose the way ASTs are loaded during "
+                                       "CTU analysis. Mode 'from-pch' "
+                                       "generates PCH format serialized ASTs "
+                                       "during the 'collect' phase. Mode "
+                                       "'parsed-on-demand' only generates the "
+                                       "invocations needed to parse the ASTs. "
+                                       "Mode 'from-pch' can use significant "
+                                       "disk-space for the serialized ASTs, "
+                                       "while mode 'parsed-on-demand' can "
+                                       "incur some runtime overhead in the "
```
Comment (gyorb) [inline analyzer/codechecker_analyzer/cmd/analyze.py:485]: runtime memory or CPU or both overhead?

File: analyzer/codechecker_analyzer/pre_analysis_manager.py [hunk 13]
```diff
@@ -114,8 +111,22 @@ def pre_analyze(params):
                 ctu_triple_arch.get_triple_arch(action, action.source,
                                                 clangsa_config,
                                                 analyzer_environment)
-            ctu_manager.generate_ast(triple_arch, action, action.source,
-                                     clangsa_config, analyzer_environment)
+
+            # TODO: reorganize the various ctu modes parameters
+            # Dump-based analysis requires serialized ASTs.
+            if clangsa_config.ctu_on_demand:
+                ctu_manager.generate_invocation_list(triple_arch, action,
```
Comment (gyorb) [inline analyzer/codechecker_analyzer/pre_analysis_manager.py:118]: is the invocation list different from the file generated by the extdef-mapping tool?

File: analyzer/codechecker_analyzer/pre_analysis_manager.py [hunk 14]
```diff
@@ -203,9 +214,9 @@ def signal_handler(signum, frame):
     # Postprocessing the pre analysis results.
     if ctu_data:
         ctu_manager.merge_ctu_func_maps(
-                ctu_data.get('ctu_dir'),
-                ctu_data.get('ctu_func_map_file'),
-                ctu_data.get('ctu_temp_fnmap_folder'))
+            ctu_data.get('ctu_dir'),
+            ctu_data.get('ctu_func_map_file'),
```
Comment (gyorb) [inline analyzer/codechecker_analyzer/pre_analysis_manager.py:218]: is the file name still ctu_func_map_file, shouldn't that be extdef-mapping file or something like that?

File: analyzer/codechecker_analyzer/cmd/analyze.py (hunk 11 above)
Comment (gamesh411) [reply analyzer/codechecker_analyzer/cmd/analyze.py:471]: This seems tricky to get right semantically. @martong @csordasmarton @bruntib  what do you think about `ctu-ast-mode`?

File: analyzer/codechecker_analyzer/cmd/analyze.py (hunk 12 above)
Comment (gamesh411) [reply analyzer/codechecker_analyzer/cmd/analyze.py:485]: only CPU, i will mention it in the comments

File: analyzer/codechecker_analyzer/pre_analysis_manager.py (hunk 13 above)
Comment (gamesh411) [reply analyzer/codechecker_analyzer/pre_analysis_manager.py:118]: Yes, the invocation list file is a mapping from the source file to the clang invocation, the extdef-mapping file is a mapping from usr-names to source files (or ast dumps in case of dump-based analysis).

File: analyzer/codechecker_analyzer/pre_analysis_manager.py (hunk 14 above)
Comment (gamesh411) [reply analyzer/codechecker_analyzer/pre_analysis_manager.py:218]: That would make sense, however I think this change is just left here unintentionally (maybe auto-formatted accidentally :P). I will remove this part of the diff.

File: analyzer/codechecker_analyzer/analyzers/clangsa/ctu_manager.py (hunk 9 above)
Comment (gamesh411) [reply analyzer/codechecker_analyzer/analyzers/clangsa/ctu_manager.py:120]: yeeey, cant wait for the shiny new YAML library :D ! 👍

File: analyzer/codechecker_analyzer/analyzers/clangsa/ctu_manager.py (hunk 10 above)
Comment (gamesh411) [reply analyzer/codechecker_analyzer/analyzers/clangsa/ctu_manager.py:126]: I always forget these...

**Comment by gyorb** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- analyzer/tests/functional/ctu_failure/test_ctu_failure.py  1
- analyzer/codechecker_analyzer/analyzers/clangsa/analyzer.py  1
         

Clones added
============
- analyzer/tests/functional/ctu_failure/test_ctu_failure.py  25
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/CodeChecker/codechecker/pullRequest?prid=3892752&bid=13556301)

File: analyzer/codechecker_analyzer/cmd/analyze.py (hunk 11 above)
Comment (csordasmarton) [reply analyzer/codechecker_analyzer/cmd/analyze.py:471]: I vote on the `ctu-ast-mode` name. It is more expressive.

File: docs/analyzer/user_guide.md [hunk 15]
```diff
@@ -1243,9 +1244,18 @@ cross translation unit analysis arguments:
                         analysis, using already available extra files in
                         '<OUTPUT_DIR>/ctu-dir'. (These files will not be
                         cleaned up in this mode.)
-  --ctu-on-the-fly      If specified, the 'collect' phase will not create the
-                        extra AST dumps, but rather analysis will be run with
-                        an in-memory recompilation of the source files.
+  --ctu-ast-creation {from-pch, parsed-on-demand}
+                        Choose the way ASTs are loaded during CTU analysis.
```
Comment (csordasmarton) [inline docs/analyzer/user_guide.md:1248]: This help message is different from the help message written in the source code.

File: docs/analyzer/user_guide.md [hunk 16]
```diff
@@ -645,6 +645,7 @@ usage: CodeChecker analyze [-h] [-j JOBS]
                            [--tidyargs TIDY_ARGS_CFG_FILE]
                            [--tidy-config TIDY_CONFIG] [--timeout TIMEOUT]
                            [--ctu | --ctu-collect | --ctu-analyze]
+                           [--ctu-ast-creation {from-pch, parsed-on-demand}]
```
Comment (csordasmarton) [inline docs/analyzer/user_guide.md:648]: CodeChecker check command also has this option. Please extend the documentation for the check command too.

File: analyzer/tests/functional/ctu_failure/test_ctu_failure.py [hunk 17]
```diff
@@ -229,16 +360,48 @@ def test_ctu_fallback(self):
...
         self.assertIn("main.c", failed_files[0])
 
-    def __do_ctu_all(self, reparse, extra_args=None):
+    def test_ctu_on_demand_fallback(self):
+        """ In case of ctu failure the non ctu analysis will be triggered when
+        using on-demand-mode.
+        """
+        self.__set_up_test_dir('ctu_failure')
+        if not self.ctu_capable:
+            self.skipTest(NO_CTU_MESSAGE)
```
Comment (csordasmarton) [inline analyzer/tests/functional/ctu_failure/test_ctu_failure.py:369]: Cannot we use the `@unittest.skip` decorator to skip this test case based on the given conditions?

For more info see: https://docs.python.org/3/library/unittest.html

File: analyzer/tests/functional/ctu_failure/test_ctu_failure.py [hunk 18]
```diff
@@ -146,6 +167,60 @@ def check_source_in_archive(source_in_archive):
             check_source_in_archive("main.c")
             check_source_in_archive("lib.c")
 
+    def test_ctu_on_demand_failure_zip(self):
+        """ Test the failure zip contains the source of imported TU when using
+        on-demand mode.
+        """
+        self.__set_up_test_dir('ctu_failure')
```
Comment (csordasmarton) [inline analyzer/tests/functional/ctu_failure/test_ctu_failure.py:174]: We use the same directory name as in the `test_ctu_on_demand_logs_ast_import` function. Isn't it a problem?

File: analyzer/tests/functional/ctu_failure/test_ctu_failure.py [hunk 19]
```diff
@@ -200,6 +275,62 @@ def check_source_in_archive(source_in_archive):
...
+        Test the failure zip contains the source of imported TU and all the
+        headers on which the TU depends when using on-demand mode.
+        """
+        self.__set_up_test_dir('ctu_failure_with_headers')
+        if not self.ctu_capable:
+            self.skipTest(NO_CTU_MESSAGE)
+        if not self.ctu_has_analyzer_display_ctu_progress:
+            self.skipTest(NO_DISPLAY_CTU_PROGRESS)
+        if not self.ctu_on_demand_capable:
+            self.skipTest(NO_CTU_ON_DEMAND_MESSAGE)
```
Comment (csordasmarton) [inline analyzer/tests/functional/ctu_failure/test_ctu_failure.py:289]: This code is duplicated in multiple places. Can we create a separate helper function for it?

**Comment by gyorb** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- analyzer/tests/functional/ctu_failure/test_ctu_failure.py  1
- analyzer/codechecker_analyzer/analyzers/clangsa/analyzer.py  1
         

Clones added
============
- analyzer/tests/functional/ctu_failure/test_ctu_failure.py  25
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/CodeChecker/codechecker/pullRequest?prid=3892752&bid=13556301)

File: analyzer/codechecker_analyzer/analyzers/clangsa/analyzer.py [hunk 20]
```diff
@@ -336,7 +379,9 @@ def construct_config_handler(cls, args, context):
         if 'ctu_phases' in args:
             handler.ctu_dir = os.path.join(args.output_path,
                                            args.ctu_dir)
-
+            handler.ctu_on_demand = \
+                'ctu_ast_creation' in args and \
+                args.ctu_ast_loading_mode == 'parsed-on-demand'
```
Comment (csordasmarton) [inline analyzer/codechecker_analyzer/analyzers/clangsa/analyzer.py:384]:
```suggestion
                args.ctu_ast_creation == 'parsed-on-demand'
```

Comment (gamesh411) [reply analyzer/codechecker_analyzer/analyzers/clangsa/analyzer.py]: Thanks for catching this, I have updated both occurences with ctu_ast_mode now.

**Comment by gyorb** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- analyzer/tests/functional/ctu_failure/test_ctu_failure.py  1
- analyzer/codechecker_analyzer/analyzers/clangsa/analyzer.py  1
         

Clones added
============
- analyzer/tests/functional/ctu_failure/test_ctu_failure.py  25
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/CodeChecker/codechecker/pullRequest?prid=3892752&bid=13556301)

File: analyzer/tests/functional/ctu_failure/test_ctu_failure.py (hunk 18 above)
Comment (gamesh411) [reply analyzer/tests/functional/ctu_failure/test_ctu_failure.py:174]: This might be a problem. To be safe I will use another prefix/postfix for this test.

**Comment by gyorb** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- analyzer/tests/functional/ctu_failure/test_ctu_failure.py  1
- analyzer/codechecker_analyzer/analyzers/clangsa/analyzer.py  1
         

Clones added
============
- analyzer/tests/functional/ctu_failure/test_ctu_failure.py  25
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/CodeChecker/codechecker/pullRequest?prid=3892752&bid=13556301)

File: analyzer/tests/functional/ctu_failure/test_ctu_failure.py (hunk 17 above)
Comment (gamesh411) [reply analyzer/tests/functional/ctu_failure/test_ctu_failure.py]: Decorators rule, thanks for pointing them out. I have implemented the test skipping that way.

**Comment by gyorb** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- analyzer/codechecker_analyzer/analyzers/clangsa/analyzer.py  1
         

Clones added
============
- analyzer/tests/functional/ctu/test_ctu.py  1
- analyzer/tests/functional/ctu_failure/test_ctu_failure.py  13
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/CodeChecker/codechecker/pullRequest?prid=3892752&bid=13556301)

**Comment by gyorb** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- analyzer/codechecker_analyzer/analyzers/clangsa/analyzer.py  1
         

Clones added
============
- analyzer/tests/functional/ctu/test_ctu.py  1
- analyzer/tests/functional/ctu_failure/test_ctu_failure.py  13
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/CodeChecker/codechecker/pullRequest?prid=3892752&bid=13556301)

**Comment by gyorb** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- analyzer/codechecker_analyzer/analyzers/clangsa/analyzer.py  1
         

Clones added
============
- analyzer/tests/functional/ctu_failure/test_ctu_failure.py  13
- analyzer/tests/functional/ctu/test_ctu.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/CodeChecker/codechecker/pullRequest?prid=3892752&bid=13556301)

**Comment by gyorb** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- analyzer/codechecker_analyzer/analyzers/clangsa/analyzer.py  1
         

Clones added
============
- analyzer/tests/functional/ctu_failure/test_ctu_failure.py  13
- analyzer/tests/functional/ctu/test_ctu.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/CodeChecker/codechecker/pullRequest?prid=3892752&bid=13556301)

**Comment by gyorb** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- analyzer/codechecker_analyzer/analyzers/clangsa/analyzer.py  1
         

Clones added
============
- analyzer/tests/functional/ctu_failure/test_ctu_failure.py  13
- analyzer/tests/functional/ctu/test_ctu.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/CodeChecker/codechecker/pullRequest?prid=3892752&bid=13556301)

**Comment by gyorb** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- analyzer/codechecker_analyzer/analyzers/clangsa/analyzer.py  1
         

Clones added
============
- analyzer/tests/functional/ctu/test_ctu.py  1
- analyzer/tests/functional/ctu_failure/test_ctu_failure.py  13
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/CodeChecker/codechecker/pullRequest?prid=3892752&bid=13556301)
